google/stackdriver_monitoring/prometheus_service.py

```diff
@@ -0,0 +1,187 @@
+    # do to hook it into ours so we'll let the client library use its server
+    # for now.
+    parser.add_argument(
+        '--prometheus_port', default=8003, type=int,
+        help='Port for Prometheus HTTP Server')
+    parser.add_argument(
+        '--prometheus_add_source_metalabels', default=True,
+        action='store_true',
+        help='Add Spinnaker job/instance labels for prometheus.')
+
+  def __init__(self, options):
+    self.__service_endpoints = spectator_client.determine_service_endpoints(
+        options)
+    self.__spectator = spectator_client.SpectatorClient(options)
+    self.__add_metalabels = options.get('prometheus_add_source_metalabels',
+                                        True)
+
+    REGISTRY.register(self)
+    port = options['prometheus_port']
+    logging.info('Starting prometheus http server on port %d' % port)
+    start_http_server(port)
+
+  def __collect_instance_info(
+        self, service, name,
+        instance, metric_metadata, service_metadata, service_to_name_to_info):
+    """Creates a post payload for a DataDog time series data point.
+
+       See http://docs.datadoghq.com/api/?lang=python#metrics-post.
+
+    Args:
+      service: [string] The name of the service that the metric is from.
+      name: [string] The name of the metric coming from the service.
+      instance: [dict] The spectator entry for a specific metric value
+         for a specific tag binding instance that we're going to append.
+      metric_metadata: [dict] The spectator JSON object for the metric
+         is used to get the kind and possibly other metadata.
+      service_to_name_to_info: [dict] A dictionary keyed by service to
+        A dictionary maping metric names to MetricInfo being built.
+    """
+    # In practice this converts a Spinnaker Timer into either
+    # <name>__count or <name>__totalTime and removes the "statistic" tag.
+    name, tags = spectator_client.normalize_name_and_tags(
+        name, instance, metric_metadata)
+    if tags is None:
+      return  # ignore metrics that had no tags because these are bogus.
+
+    record = InstanceRecord(service,
+                            '{0}:{1}'.format(service_metadata['__host'],
+                                             service_metadata['__port']),
+                            instance)
+
+    name_to_info = service_to_name_to_info.get(service)
+    if name_to_info is None:
+      name_to_info = {}
+      service_to_name_to_info[service] = name_to_info
+
+    tag_names = set([tag['key'] for tag in tags])
+    info = name_to_info.get(name)
+    if info is None:
+      info = MetricInfo(metric_metadata['kind'], tag_names, [record])
+      name_to_info[name] = info
+      return
+
+    info.records.append(record)
+    info.tags.update(tag_names)
+
+    # In practice this converts a Spinnaker Timer into either
+    # <name>__count or <name>__totalTime and removes the "statistic" tag.
+    name, tags = spectator_client.normalize_name_and_tags(
```

Comment:
You call this exact block of code above, either factor out into a function or remove if it's duplicated.